tsl/src/chunk_copy.c

```diff
@@ -0,0 +1,581 @@
+				   " WITH (create_slot = false, enabled = false)",
+				   NameStr(cc->operation_id),
+				   connection_string,
+				   NameStr(cc->operation_id));
+	ts_dist_cmd_run_on_data_nodes(cmd, list_make1((char *) cc->dst_node), false);
+}
+
+static void
+chunk_copy_stage_sync_start(ChunkCopy *cc)
+{
+	const char *cmd;
+
+	/* Start data transfer on the destination node */
+	cmd = psprintf("ALTER SUBSCRIPTION %s ENABLE", NameStr(cc->operation_id));
+	ts_dist_cmd_run_on_data_nodes(cmd, list_make1((char *) cc->dst_node), false);
+}
+
+static void
+chunk_copy_stage_sync(ChunkCopy *cc)
+{
+	const char *cmd;
+
+	/* Wait until data transfer finishes in its own transaction */
+	cmd = psprintf("CALL _timescaledb_internal.wait_subscription_sync(%s, %s)",
+				   quote_literal_cstr(NameStr(cc->chunk->fd.schema_name)),
+				   quote_literal_cstr(NameStr(cc->chunk->fd.table_name)));
+
+	ts_dist_cmd_run_on_data_nodes(cmd, list_make1((char *) cc->dst_node), false);
+}
+
+static void
+chunk_copy_stage_drop_subscription(ChunkCopy *cc)
+{
+	const char *cmd;
+
+	cmd = psprintf("DROP SUBSCRIPTION %s", NameStr(cc->operation_id));
+	ts_dist_cmd_run_on_data_nodes(cmd, list_make1((char *) cc->dst_node), false);
+}
+
+static void
+chunk_copy_stage_drop_publication(ChunkCopy *cc)
+{
+	const char *cmd;
+
+	cmd = psprintf("DROP PUBLICATION %s", NameStr(cc->operation_id));
+	ts_dist_cmd_run_on_data_nodes(cmd, list_make1((char *) cc->src_node), false);
+}
+
+static void
+chunk_copy_stage_attach_chunk(ChunkCopy *cc)
+{
+	Cache *hcache;
+	Hypertable *ht;
+	ChunkDataNode *chunk_data_node;
+	const char *remote_chunk_name;
+	Chunk *chunk = cc->chunk;
+
+	ht = ts_hypertable_cache_get_cache_and_entry(chunk->hypertable_relid, CACHE_FLAG_NONE, &hcache);
+
+	/* Check that the hypertable is already attached to this data node */
+	data_node_hypertable_get_by_node_name(ht, cc->dst_server->servername, true);
+
+	chunk_data_node = palloc0(sizeof(ChunkDataNode));
+
+	chunk_data_node->fd.chunk_id = chunk->fd.id;
+	chunk_data_node->fd.node_chunk_id = -1; /* below API will fill it up */
+	namestrcpy(&chunk_data_node->fd.node_name, cc->dst_server->servername);
+	chunk_data_node->foreign_server_oid = cc->dst_server->serverid;
+
+	remote_chunk_name = psprintf("%s.%s",
+								 quote_identifier(chunk->fd.schema_name.data),
+								 quote_identifier(chunk->fd.table_name.data));
+
+	chunk_api_create_on_data_nodes(chunk, ht, remote_chunk_name, list_make1(chunk_data_node));
+
+	/* All ok, update the AN chunk metadata to add this data node to it */
+	chunk->data_nodes = lappend(chunk->data_nodes, chunk_data_node);
+
+	/* persist this association in the metadata */
+	ts_chunk_data_node_insert(chunk_data_node);
+
+	ts_cache_release(hcache);
+}
+
+static void
+chunk_copy_stage_delete_chunk(ChunkCopy *cc)
+{
+	if (!cc->delete_on_src_node)
+		return;
+
+	chunk_api_call_chunk_drop_replica(cc->chunk, cc->src_node, cc->src_server->serverid);
+}
+
+static ChunkCopyStage copy_stages[] = {
```

Comment:
`👍 for this.

But I think this should be const, right? Or is it not possible for some reason?

```suggestion
static const ChunkCopyStage copy_stages[] = {
```